### PR TITLE
Gearama/profile debug

### PR DESCRIPTION
**Pull request by gearama**
Enable the  passing of the debug option to the setup method for languages (it was missing) also enable running valgrind tool for profiling for cpp . Should i do the install of the tool here or in the pipeline? i tend for the pipeline since i don't know exactly what linux i'm runing and what package tool it it using. 
Also added a profiler options param. valgrind has several profilers that can be selected through command line, thus i need this param to be able to execute properly. 

Added tests for the method creating the command lines to be executed to build. Since the Util methods are static in a static class they cannot be mocked, thus i had to work around it to get the parameters of the method calls to be able to compare them with expected values.

### CODE REVIEW

**Review by mikeharder — CHANGES_REQUESTED** [review]

File: tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs [hunk 1]
```diff
@@ -42,8 +43,8 @@ public class Cpp : LanguageBase
             await UpdatePackageVersions(packageVersions);
 
             // Windows and Linux require different arguments to build Release config
-            var additionalGenerateArguments = Util.IsWindows ? "-DDISABLE_AZURE_CORE_OPENTELEMETRY=ON" : "-DCMAKE_BUILD_TYPE=Release";
-            var additionalBuildArguments = Util.IsWindows ? "--config MinSizeRel" : String.Empty;
+            var additionalGenerateArguments = Util.IsWindows ? "-DDISABLE_AZURE_CORE_OPENTELEMETRY=ON" : (debug ? "-DCMAKE_BUILD_TYPE=Debug" : "-DCMAKE_BUILD_TYPE=Release");
```
[the remark this answers is not in the document]
Comment (mikeharder) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs:46]: The more I think about it, the more I like keeping the `--debug` option to the perf automation, and flowing it through to the languages to enable whatever debugging they see fit.  The idea is if the perf automation or tests are failing, and you want to get more information but don't care about the perf numbers, add the `--debug` option.

We could add a top-level parameter to our pipelines to enable this, but for now I think I would just tell people to put `--debug` in the existing `AdditionalArguments` textbox.

[the remark this answers is not in the document]
Comment (gearama) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs:46]: debgu helps with getting stack traces for memory leaks or other profiling issues. once we identify an issue , running the debuig version should allow for more detailed information as to where things happen, which callstack.

File: tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/LanguageBase.cs [hunk 2]
```diff
@@ -9,6 +9,8 @@ public abstract class LanguageBase : ILanguage
     {
         protected abstract Language Language { get; }
 
+        protected string ProfileDirectory => Path.GetFullPath(Path.Combine(WorkingDirectory, Language + "-profile"));
```
[the remark this answers is not in the document]
Comment (gearama) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/LanguageBase.cs:12]: done

File: tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/ILanguage.cs [hunk 3]
```diff
@@ -16,7 +16,8 @@ public interface ILanguage
             string project,
             string languageVersion,
             string primaryPackage,
-            IDictionary<string, string> packageVersions);
+            IDictionary<string, string> packageVersions,
+            bool debug = false);
```
[the remark this answers is not in the document]
Comment (gearama) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/ILanguage.cs:20]: done

File: tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Program.cs [hunk 4]
```diff
@@ -354,7 +364,7 @@ private static async Task Run(Options options)
                                         test.Class,
                                         allArguments,
                                         options.Profile,
-                                        context
+                                        contextWrapper.Count() != 0 ? contextWrapper : context
```
[the remark this answers is not in the document]
Comment (gearama) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Program.cs:367]: done

File: tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs [hunk 5]
```diff
@@ -68,11 +69,46 @@ public class Cpp : LanguageBase
             string testName,
             string arguments,
             bool profile,
-            object context)
+            object context
+            )
```
[the remark this answers is not in the document]
Comment (gearama) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs:73]: done

File: tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs [hunk 6]
```diff
@@ -68,11 +69,46 @@ public class Cpp : LanguageBase
...
+            }
+
+            var profiledExe = perfExe;
+
+            if (profile)
+            {
+                if (Util.IsWindows)
+                {
+                    Console.WriteLine("Profiling available on linux alone at the moment");
+                    profiling = false;
```
[the remark this answers is not in the document]
Comment (gearama) [reply tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Cpp.cs:95]: done

**Comment by mikeharder** [comment]
@gearama: The code changes look good to me, however before merging this PR I'd like to verify it works end-to-end, and I have a few questions related to this.

1. Should profiling work correctly if `ProfilerOptions` is not set?  If it currently does not, can we define a reasonable set of defaults so it does? 
2. Should we add a top-level `ProfilerOptions` to the C++ pipelines (and the `tools - perf-automation` pipeline)?  Or were you thinking users would set these via the `AdditionalArguments` textbox that already exists?

For installing the `valgrind` tool, I would add this to the pipeline YML in both the tools repo (for the `tools - perf-automation` pipeline) and the C++ repo (for your service pipelines).  For an example, see how .NET is installing their profiling tool:

https://github.com/Azure/azure-sdk-tools/pull/5572/files#diff-db1428e1537bee9d1dbbf6d610c0bc30233911688ebe327425ab73c96536e39dR163

https://github.com/Azure/azure-sdk-for-net/pull/34554/files#diff-fe881f2374762f98475e30354d6f9c3e23c2265394a0b55e50a9e6c238532384R41

As part of the C++ PR to install `valgrind`, you can also add the `Profile` parameter (and optionally the `ProfilerOptions` parameter) to your service pipelines (like .NET did).

**Comment by gearama** [comment]
Should profiling work correctly if ProfilerOptions is not set? If it currently does not, can we define a reasonable set of defaults so it does?
[[gearama]] yes, in the case of c++ it defaults to memory leaks checks. 

Should we add a top-level ProfilerOptions to the C++ pipelines (and the tools - perf-automation pipeline)? Or were you thinking users would set these via the AdditionalArguments textbox that already exists?

[[gearama]] i would make it a separate field if needed, if it works as aditional params i would not bother adding it as a separata field.
For installing the valgrind tool, I would add this to the pipeline YML in both the tools repo (for the tools - perf-automation pipeline) and the C++ repo (for your service pipelines). For an example, see how .NET is installing their profiling tool:

**Comment by gearama** [comment]
https://github.com/Azure/azure-sdk-for-cpp/pull/4398

**Comment by mikeharder** [comment]
I am able to run this PR with no errors, and I can verify valgrind was run, but the generated profile zip is empty, because no profile output files were generated under the `profile` directory.  You can obtain the path to this directory using this util method:

https://github.com/Azure/azure-sdk-tools/blob/da1672215831923cc94f0e17a9f55d3a04689e55/tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/Util.cs#L120-L123

You will need to generate an output filename, example in JS here:

https://github.com/Azure/azure-sdk-tools/blob/da1672215831923cc94f0e17a9f55d3a04689e55/tools/perf-automation/Azure.Sdk.Tools.PerfAutomation/JavaScript.cs#L150-L152

Then either pass this filename to `valgrind` so it knows where to write the profile, or move the file generated by `valgrind` to this path.

**Review by mikeharder — CHANGES_REQUESTED** [review]
If a profile output file should be generated by default, this PR needs additional changes.

**Comment by gearama** [comment]
by default valgrind outputs to stderr , i prefer to leave it li it is for the moment untill the need for a file presents itself. personally i don't see the need for one,all the data is in the log. 
IMHO a log file is more headache then it's worth, i have to hunt down some file in who knows what location, then i have to figure out whom to beg for access in case i don't have it(most likely)  just to get data that is already in the log :)  . does not seem worth the hastle when i can just scroll the log.

**Comment by mikeharder** [comment]
> by default valgrind outputs to stderr , i prefer to leave it as it is for the moment until the need for a file presents itself. personally i don't see the need for one, all the data is in the log. IMHO a log file is more headache then it's worth, i have to hunt down some file in who knows what location, to which i may or may not have access, then i have to figure out whom to beg for access in case i don't have it(most likely) just to get data that is already in the log :) . does not seem worth the hastle when i can just scroll the log. Besides the log file format is known by everyone , thus makes it strait forward for people to understand. Also i prefer to see the output as part of the whole execution log, as it can be corelated with the rest of the logs for a big picture view.

If the data you need is in the console log, I agree publishing to an artifact file is optional.  Most profilers in other languages output a binary file that needs to be loaded in a tool to view the data, so these are published in an artifact zip file, since there would be no way to view the data in the console log.

**Review by mikeharder — APPROVED** [review]

**Comment by mikeharder** [comment]
This LGTM, we can merge this now and add more features to cpp profiling in future PRs.

**Pull request merged**